submissions/Chikus/port-sniffer/sniffer.js

```diff
@@ -0,0 +1,123 @@
+const Net = require('net');
+
+function sniffer(arg) {
+  if (arg.length > 6) {
+    throw new Error(
+      'Bad usage to know the manual type: node sniffer.js --help'
+    );
+  }
+  const help = `
+TCP sniffer parameters:
+    --host     Mandatory parameter which define the host address for a TCP
+               port, It can be specified as IP address or URL.
+    --ports    Optional parameter which specifies port range, from min (1) to
+               max (65536) with a separator '-' between them, if this is not specified will
+               take the min and max values.
+
+Examples: node sniffer.js --host google.com --ports 100-2345
+          node sniffer.js --ports 1-1500 --host www.google.com
+          node sniffer.js --host 171.217.3.110
+          node sniffer.js --ports 70-80 --host 171.217.3.110
+          node sniffer.js --host www.google.com
+`;
+
+  function checkServer(host, port) {
+    return new Promise(function prom(resolve) {
+      const socket = Net.createConnection(port, host);
+      socket.setTimeout(300);
+      socket.on('timeout', function timer() {
+        socket.destroy();
+        resolve(null);
+      });
+      socket.on('connect', function success() {
+        process.stdout.write('.');
+        socket.destroy();
+        resolve(port);
+      });
+      socket.on('error', function fail() {
+        throw new Error('Error in Socket, wrong host \n');
+      });
+    }).catch(e => e);
+  }
+
+  async function runPorts(host, lowerPort, higherPort) {
+    const portList = [];
+    for (let i = lowerPort; i <= higherPort; i += 1) {
+      /* eslint-disable */
+      const validPort = await checkServer(host, i);
+      /* eslint-enable */
+      if (validPort) {
+        portList.push(validPort);
+      }
+    }
+    if (portList.length) {
+      process.stdout.write(`\n${portList.join()} ports are opened \n`);
+      return process.exit(0);
+    }
+    process.stdout.write('No ports were found \n');
+    return process.exit(0);
+  }
+
+  function checkPorts(userPorts) {
+    if (userPorts) {
+      const ports = userPorts.split('-').map(elem => Number(elem));
+      if (
+        ports[0] > ports[1] ||
+        ports[0] < 0 ||
+        ports[1] > 65535 ||
+        !ports[1]
+      ) {
+        throw new Error(' Check your ports, range from 1-65535\n');
+      }
+      return [ports[0], ports[1]];
+    }
+    throw new Error('Ports no defined \n');
+  }
+
+  function checkHost(host) {
+    if (host) {
+      const domains = host.split('.');
+      if (host.replace(/[^.]/g, '').length === 3) {
+        if (
+          domains.every(x => Number.isNaN(x)) ||
+          domains.every(x => !Number.isNaN(x))
+        ) {
+          return host;
+        }
+      } else if (host.replace(/[^.]/g, '').length <= 2) {
+        if (domains.every(x => !Number.isNaN(x))) {
+          return host;
+        }
+      }
+    }
+    return null;
+  }
+
+  switch (arg[2]) {
+    case '--ports':
```

Comment:
Instead of using explicit switch case which will grow very rapidly if you gonna add more params to your program, it's better to create more universal solution to parse arguments. Maybe something like this (simplified):
```js
function parseArguments(args) {
   const parsedArgs = args.slice(2).reduce((acc, arg) => {
         if (arg.startsWith('--') && !acc.lastKey) {
              acc.lastKey = arg.replace('--', '')
         } else if (acc.lastKey) {
              acc.result = {...acc.result, [acc.lastKey]: arg}
              acc.lastKey = null
         } else {
               throw new Error(`Unexpected argument '${arg}'`)
         }
         return acc
    }, {result: {}, lastKey: null}).result
    if (parsedArgs.lastKey) throw new Error(`Unexpected key '${acc.lastKey}' without value`)
    return parsedArgs.result
}
```

Reply:
```js
const args = ['node',
  '/home/index.js',
  '--one',
  'two',
  '--three',
  'four']
parseArguments(args)
// {one: "two", three: "four"}
```

Reply:
hehe you gave me the solution, thanks, I will make the changes.

Reply:
This was a very challenging function, this reduce is kind of crazy, but I like it and it has a bunch of functionalities. I hope in the future I can use it again :D

Reply:
In real world apps you might be using external package for that. (like `yargs`) My implementation is very straightforward.